### PR TITLE
generate standardized documentation (GSI-143)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
```
- .readme_template.md
- setup.cfg
- config_schema.json
- openapi.yaml
- .description.md
- .design.md
```

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -16,7 +16,7 @@
 [metadata]
 name = dcs
 version = attr: dcs.__version__
-description = dcs - a GA4GH drs compliant service for delivering files from S3
+description = Download Controller Service - a GA4GH DRS-compliant service for delivering files encrypted according to the GA4GH Crypt4GH standard from S3.
```
Comment (mephenor) [inline setup.cfg:19]:
```suggestion
description = Download Controller Service - a GA4GH DRS-compliant service for delivering files from S3 encrypted according to the GA4GH Crypt4GH standard.
```

File: .description.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<!-- Please provide a short overview of the features of this service.-->
+
+This service implements the
+[GA4GH DRS](https://github.com/ga4gh/data-repository-service-schemas) v1.0.0 for
+serving files that where encrypted according to the
+[GA4GH Crypt4GH](https://www.ga4gh.org/news/crypt4gh-a-secure-method-for-sharing-human-genetic-data/)
+from S3-compatible object storages.
+
+Thereby, only the `GET /objects/{object_id}` is implemented. It always return
```
Comment (mephenor) [inline .description.md:9]:
```suggestion
Thereby, only the `GET /objects/{object_id}` is implemented. It always returns
```

File: .description.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+an access_method for accessing the object via S3. This makes the second endpoint
```
Comment (mephenor) [inline .description.md:10]:
```suggestion
an access_method for the object via S3. This makes the second endpoint
```

File: .description.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+`GET /objects/{object_id}/access/{access_id}` that
+is contained in the DRS spec unnecessary. For more details see the OpenAPI spec
+described below.
+
+For authorization, a JWT token is expected via Bearer Authentication that has a format
+described [here](./dcs/core/auth_policies.py).
+
+All files that can be requested are registered in a MongoDB database owned and
+controlled by this service. Registration of new events happens through an Kafka event.
```
Comment (mephenor) [inline .description.md:19]:
```suggestion
controlled by this service. Registration of new events happens through a Kafka event.
```

File: .description.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+It serves pre-signed URLs to S3 objects located in a single so-called outbox bucket.
+If the file is not already in the bucket upon when the user calls the object endpoint,
```
Comment (mephenor) [inline .description.md:22]:
```suggestion
If the file is not already in the bucket when the user calls the object endpoint,
```

File: .description.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+an event is published to request staging the file to the outbox. The staging has to
+be carried out by a different service.
+
+For more details on the events consumed and produced by this service, see the
+configuration.
+
+The DRS object endpoint serves files in an encrypted fashion as described by the
+Crypt4GH standard but without the evelope. A user-specific envelope can be requested
```
Comment (mephenor) [inline .description.md:30]:
```suggestion
Crypt4GH standard, but without the evelope. A user-specific envelope can be requested
```

File: .description.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from the `GET /objects/{object_id}/envelopes` endpoint. The actual envelope creation
+is delecated to another service via a RESTful call. Please see the configuration for
```
Comment (mephenor) [inline .description.md:32]:
```suggestion
is delegated to another service via a RESTful call. Please see the configuration for
```

**Review by mephenor — APPROVED** [review]
Some typos/phrasing issues, rest looks good.

File: dcs/adapters/outbound/http/api_calls.py [hunk 8]
```diff
@@ -24,7 +26,11 @@
 def call_ekss_api(*, secret_id: str, receiver_public_key: str, api_base: str) -> str:
     """Calls EKS to get an envelope for an encrypted file, using the receivers
     public key as well as the id of the file secret."""
-    api_url = f"{api_base}/secrets/{secret_id}/envelopes/{receiver_public_key}"
+
+    receiver_public_key_base64 = base64.urlsafe_b64encode(
+        receiver_public_key.encode()
+    ).decode()
```
Comment (mephenor) [inline dcs/adapters/outbound/http/api_calls.py:32]:
```suggestion
    receiver_public_key_base64 = base64.urlsafe_b64encode(
        base64.b64decode(receiver_public_key)
    ).decode()
```
Current solution would reencode the already encoded key

**Review by mephenor — CHANGES_REQUESTED** [review]

File: dcs/adapters/outbound/http/api_calls.py (hunk 8 above)
Comment (KerstenBreuer) [reply dcs/adapters/outbound/http/api_calls.py:32]: So the receiver_public_key is already b64 encoded but not in a urlsafe way?

File: dcs/adapters/outbound/http/api_calls.py [hunk 9]
```diff
@@ -24,7 +26,11 @@
...
+    api_url = f"{api_base}/secrets/{secret_id}/envelopes/'{receiver_public_key_base64}'"
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply dcs/adapters/outbound/http/api_calls.py]: You are right, this is not necessary.

File: dcs/adapters/outbound/http/api_calls.py (hunk 8 above)
Comment (mephenor) [reply dcs/adapters/outbound/http/api_calls.py:32]: pubkey comes from the work order token which should include this as base64 encoded string

**Review by mephenor — APPROVED** [review]

File: .description.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+[GA4GH Crypt4GH](https://www.ga4gh.org/news/crypt4gh-a-secure-method-for-sharing-human-genetic-data/)
+from S3-compatible object storages.
+
+Thereby, only the `GET /objects/{object_id}` is implemented. It always returns
+an access_method for the object via S3. This makes the second endpoint
+`GET /objects/{object_id}/access/{access_id}` that
+is contained in the DRS spec unnecessary. For more details see the OpenAPI spec
+described below.
+
+For authorization, a JWT token is expected via Bearer Authentication that has a format
```
Comment (MoritzHahn1337) [inline .description.md:15]:
```suggestion
For authorization, a JSON web token is expected via Bearer Authentication that has a format
```

The T in JWT already stands for "token"

**Review by MoritzHahn1337 — APPROVED** [review]

**Review by MoritzHahn1337 — DISMISSED** [review]

**Review by MoritzHahn1337 — DISMISSED** [review]

**Review by mephenor — APPROVED** [review]

**Pull request merged**